File: programs/shadow-nft-standard/src/instructions/burn.rs
//! This module contains all code for the `burn` instruction,
//! which burns an NFT spl token and its metadata account
use anchor_lang::prelude::*;

use crate::{
    common::{check_ata, metadata::Metadata},
    verbose_msg,
};

#[cfg(not(feature = "use-token-2022"))]
use anchor_spl::token::{self as token_program_used, Mint, Token, TokenAccount};
#[cfg(feature = "use-token-2022")]
use {
    crate::common::token_2022::{Mint, Token2022 as Token, TokenAccount},
    spl_token_2022 as token_program_used,
};

const ZERO_DECIMALS: u8 = 0;
const BURN_ONE: u64 = 1;

/// This handler does not need to perform any checks (unless we make nfts optionally non-burnable
/// in the future) as the only thing required should be the user signature which is checked by
/// anchor's `Signer` type.
pub fn handle_burn(ctx: Context<Burn>) -> Result<()> {
    // Burn NFT
    // A check for post-mint/post-primary is not strictly required since the token burn
    // will only succeed if there exists at least one token which can only exist post-mint.
    let burn_ix = token_program_used::instruction::burn_checked(
        &token_program_used::ID,
        &ctx.accounts.owner_ata.key(),
        &ctx.accounts.asset_mint.key(),
        &ctx.accounts.owner_ata.owner,
        &[&ctx.accounts.owner.key()],
        BURN_ONE,
        ZERO_DECIMALS,
    )?;
    anchor_lang::solana_program::program::invoke(
        &burn_ix,
        &[
            ctx.accounts.owner_ata.to_account_info(),
            ctx.accounts.asset_mint.to_account_info(),
            ctx.accounts.owner.to_account_info(),
        ],
    )?;
    verbose_msg!("Burned NFT");

    // Close Token Account
    let close_ix = token_program_used::instruction::close_account(
        &token_program_used::ID,
        &ctx.accounts.owner_ata.key(),
        &ctx.accounts.owner_ata.owner,
        &ctx.accounts.owner_ata.owner,
        &[&ctx.accounts.owner_ata.owner],
    )?;
    anchor_lang::solana_program::program::invoke(
        &close_ix,
        &[
            ctx.accounts.owner_ata.to_account_info(),
            ctx.accounts.asset_mint.to_account_info(),
            ctx.accounts.owner.to_account_info(),
        ],
    )?;
    verbose_msg!("Closed Token Account");

    Ok(())
}

#[derive(Accounts)]
pub struct Burn<'info> {
    /// Metadata account to burn
    #[account(
        mut,
        close = owner,
        seeds = [
            asset_mint.key().as_ref()
        ],
        bump
    )]
    pub metadata: Account<'info, Metadata>,

    /// NFT Token Mint
    #[account(mut)]
    pub asset_mint: Account<'info, Mint>,

    /// Owner's Token Account
    #[account(
        mut,
        constraint = check_ata(&owner, &owner_ata, &asset_mint)
    )]
    pub owner_ata: Account<'info, TokenAccount>,

    /// Owner of the NFT
    #[account(mut)]
    pub owner: Signer<'info>,

    /// Token program
    pub token_program: Program<'info, Token>,

    /// System program
    pub system_program: Program<'info, System>,
}

File: js-generated-client/src/shadow-nft-standard/instructions/burn.ts
import {
  AccountMeta,
  PublicKey,
  TransactionInstruction,
} from "@solana/web3.js";

import { PROGRAM_ID } from "../programId";

export interface BurnAccounts {
  /** Metadata account to burn */
  metadata: PublicKey;
  /** NFT Token Mint */
  assetMint: PublicKey;
  /** Owner's Token Account */
  ownerAta: PublicKey;
  /** Owner of the NFT */
  owner: PublicKey;
  /** Token program */
  tokenProgram: PublicKey;
  /** System program */
  systemProgram: PublicKey;
}

/** Instruction to burn a metadata account and associated NFT. */
export function burn(
  accounts: BurnAccounts,
  programId: PublicKey = PROGRAM_ID
) {
  const keys: Array<AccountMeta> = [
    { pubkey: accounts.metadata, isSigner: false, isWritable: true },
    { pubkey: accounts.assetMint, isSigner: false, isWritable: true },
    { pubkey: accounts.ownerAta, isSigner: false, isWritable: true },
    { pubkey: accounts.owner, isSigner: true, isWritable: true },
    { pubkey: accounts.tokenProgram, isSigner: false, isWritable: false },
    { pubkey: accounts.systemProgram, isSigner: false, isWritable: false },
  ];
  const identifier = Buffer.from([116, 110, 29, 56, 107, 219, 42, 93]);
  const data = identifier;
  const ix = new TransactionInstruction({ keys, programId, data });
  return ix;
}
